backends/occa/ceed-occa-cpu-operator.cpp

```diff
@@ -0,0 +1,834 @@
+                                                const int index,
+                                                const OperatorField &opField,
+                                                const QFunctionField &qfField,
+                                                std::stringstream &dimAttribute) {
+      SimplexBasis &basis = *((SimplexBasis*) opField.basis);
+
+      dimAttribute << " @dim(";
+
+      if (qfField.evalMode == CEED_EVAL_INTERP) {
+        ss << "  const CeedScalar *" << interpVar(isInput, index);
+
+        // @dim(P, BASIS_COMPONENT_COUNT, elementCount)
+        dimAttribute << basis.P
+                     << ", " << basis.ceedComponentCount
+                     << ", elementCount";
+      }
+      else if (qfField.evalMode == CEED_EVAL_GRAD) {
+        ss << "  const CeedScalar *" << gradVar(isInput, index);
+
+        // @dim(P, BASIS_COMPONENT_COUNT, elementCount)
+        dimAttribute << basis.P
+                     << ", " << basis.ceedComponentCount
+                     << ", elementCount";
+      }
+      else if (qfField.evalMode == CEED_EVAL_WEIGHT) {
+        ss << "  const CeedScalar *" << qWeightVar(isInput, index);
+
+        // @dim(Q, elementCount)
+        dimAttribute << basis.Q << ", " << "elementCount";
+      }
+      else {
+        // Clear @dim
+        dimAttribute.str("");
+        return;
+      }
+
+      dimAttribute << ")";
+    }
+
+    void CpuOperator::addQuadArraySource(std::stringstream &ss) {
+      const int inputs = args.inputCount();
+      const int outputs = args.outputCount();
+
+      const std::string quadInput = "quadInput";
+      const std::string quadOutput = "quadOutput";
+
+      ss << "    // Store the transformed input quad values" << std::endl;
+      for (int i = 0; i < inputs; ++i) {
+        const bool isInput = true;
+        addSingleQfunctionQuadArraySource(ss, isInput, i, quadInput);
+      }
+
+      ss << std::endl
+         << "    // Store the transformed output quad values" << std::endl;
+      for (int i = 0; i < outputs; ++i) {
+        const bool isInput = false;
+        addSingleQfunctionQuadArraySource(ss, isInput, i, quadOutput);
+      }
+      ss << std::endl;
+
+      ss << std::endl
+         << "    // Store all input pointers in a single array" << std::endl;
+      addQfunctionQuadArraySource(ss, true, inputs, quadInput);
+
+      ss << std::endl
+         << "    // Store all output pointers in a single array" << std::endl;
+      addQfunctionQuadArraySource(ss, false, outputs, quadOutput);
+
+      ss << std::endl;
+    }
+
+
+    void CpuOperator::addSingleQfunctionQuadArraySource(std::stringstream &ss,
+                                                        const bool isInput,
+                                                        const int index,
+                                                        const std::string &name) {
+      // Output:
+      //   CeedScalar quadInput0[COMPONENTS][DIM][OCCA_Q];
```

Comment:
This has me concerned.

For no eval inputs/outputs, the shape should be `[comp][q]`
For interpolated inputs/outputs, the shape should be `[comp][q]`
For gradient inputs/outputs, the shape should be `[dim][comp][q]`

When I fix the interp, BP1, BP2 pass. When I fix the rest, I get segfaults.

Reply:
Thanks for checking this out, this helps a lot. I'll go ahead and update this!